airflow-core/docs/howto/deadline-alerts.rst

```diff
@@ -0,0 +1,263 @@
+    from airflow.providers.smtp.notifications.smtp import SmtpNotifier
+    from airflow.providers.standard.operators.empty import EmptyOperator
+
+    with DAG(
+        dag_id="deadline_alert_example",
+        deadline=DeadlineAlert(
+            reference=DeadlineReference.DAGRUN_QUEUED_AT,
+            interval=timedelta(minutes=15),
+            callback=SmtpNotifier(
+                to="team@example.com",
+                subject="[Alert] DAG {{ dag.dag_id }} exceeded time threshold",
+                html_content="The DAG has been running for more than 15 minutes since being queued.",
+            ),
+        ),
+    ):
+        EmptyOperator(task_id="example_task")
+
+The timeline for this example would look like this:
+
+::
+
+    |------|-----------|---------|-----------|--------|
+        Scheduled    Queued    Started    Deadline
+         00:00       00:03      00:05      00:18
+
+Using Built-in References
+-------------------------
+
+Airflow provides several built-in reference points that you can use with DeadlineAlert:
+
+``DeadlineReference.DAGRUN_QUEUED_AT``
+    Measures time from when the DagRun was queued. Useful for monitoring resource constraints.
+
+``DeadlineReference.DAGRUN_LOGICAL_DATE``
+    References when the DAG run was scheduled to start. For example, setting an interval of
+    ``timedelta(minutes=15)`` would trigger the alert if the DAG hasn't completed 15 minutes
+    after it was scheduled to start, regardless of when (or if) it actually began executing.
+    Useful for ensuring scheduled DAGs complete before their next scheduled run.
+
+``DeadlineReference.FIXED_DATETIME``
+    Specifies a fixed point in time. Useful when DAGs must complete by a specific time.
+
+Here's an example using a fixed datetime:
+
+.. code-block:: python
+
+    tomorrow_at_ten = datetime.combine(datetime.now().date() + timedelta(days=1), time(10, 0))
+
+    with DAG(
+        dag_id="fixed_deadline_alert",
+        deadline=DeadlineAlert(
+            reference=DeadlineReference.FIXED_DATETIME(tomorrow_at_ten),
+            interval=timedelta(minutes=-30),  # Alert 30 minutes before the reference.
+            callback=SmtpNotifier(
+                to="team@example.com",
+                subject="Report will be late",
+                html_content="The report will not be ready 30 minutes before the deadline.",
+            ),
+        ),
+    ):
+        EmptyOperator(task_id="example_task")
+
+The timeline for this example would look like this:
+
+::
+
+    |------|----------|---------|------------|--------|
+         Queued     Start    Deadline    Reference
+         09:15      09:17     09:30       10:00
+
+.. note::
+    Note that since the interval is a negative value, the deadline is before the reference in this case.
+
+Using Callbacks
+---------------
+
+When a deadline is exceeded, the callback is executed. You can use any async :doc:`Notifier </howto/notifications>`
```

Comment:
```suggestion
When a deadline is exceeded, the callback is executed. You can use any :doc:`Notifier </howto/notifications>`
```
Using sync or async is the goal right? Did you specifically mention async here because the sync path is still in progress? If so I'd just do something like:


```suggestion
When a deadline is exceeded, the callback is executed. You can use any :doc:`Notifier </howto/notifications>` (Note: that synchronous callback support is in development)
```

Reply:
I actually have that adjusted in a current revision.  Yes, that was my reasoning, but we're going to fast-track the sync work so I am tweaking it now.